cmd/frontend/graphqlbackend/external_account_data_resolver.go

```diff
@@ -0,0 +1,146 @@
+	data, err := publicAccountDataFromJSON(ctx, account)
+	if err != nil {
+		return nil, err
+	}
+	return &externalAccountDataResolver{
+		data: data,
+	}, nil
+}
+
+type SAMLValues struct {
+	Values *SAMLAccountData `json:"Values,omitempty"`
+}
+
+// SAML actually uses `http://schemas.xmlsoap.org...` as JSON keys for attributes
+// And we need to check multiple different attributes to find a probable username
+type SAMLAccountData struct {
+	Nickname  *SAMLAttribute `json:"nickname,omitempty"`
+	Login     *SAMLAttribute `json:"login,omitempty"`
+	Username1 *SAMLAttribute `json:"username,omitempty"`
+	Username2 *SAMLAttribute `json:"http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name,omitempty"`
+	Email1    *SAMLAttribute `json:"emailaddress,omitempty"`
+	Email2    *SAMLAttribute `json:"http://schemas.xmlsoap.org/claims/EmailAddress,omitempty"`
+	Email3    *SAMLAttribute `json:"http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress,omitempty"`
+}
+
+type SAMLAttribute struct {
+	Values []*SAMLValue `json:"Values"`
+}
+
+type SAMLValue struct {
+	Value string
+}
+
+type OpenIDConnectAccountData struct {
+	UserClaims *OpenIDUserClaims `json:"userClaims,omitempty"`
+	UserInfo   *OpenIDUserInfo   `json:"userInfo,omitempty"`
+}
+
+type OpenIDUserClaims struct {
+	Username  *string `json:"preferred_username,omitempty"`
+	GivenName *string `json:"given_name,omitempty"`
+	Name      *string `json:"name,omitempty"`
+}
+
+type OpenIDUserInfo struct {
+	Email *string `json:"email,omitempty"`
+}
+
+func publicAccountDataFromJSON(ctx context.Context, account extsvc.Account) (*extsvc.PublicAccountData, error) {
+	switch account.ServiceType {
+	case extsvc.TypeGitHub:
+		data, _, err := github.GetExternalAccountData(ctx, &account.AccountData)
+		if err != nil {
+			return nil, err
+		}
+		return &extsvc.PublicAccountData{
+			Username: data.Name,
+			Login:    data.Login,
+			URL:      data.URL,
+		}, nil
+	case extsvc.TypeGitLab:
+		data, _, err := gitlab.GetExternalAccountData(ctx, &account.AccountData)
+		if err != nil {
+			return nil, err
+		}
+		return &extsvc.PublicAccountData{
+			Username: &data.Name,
+			Login:    &data.Username,
+			URL:      &data.WebURL,
+		}, nil
+	case "saml": //TODO: define constants for SAML and OpenID Connect service types
+		data, err := encryption.DecryptJSON[SAMLValues](ctx, account.Data)
+		if err != nil {
+			return nil, err
+		}
+		v := data.Values
+		if v == nil {
+			return nil, nil
+		}
+
+		var username string
+		candidates := []*SAMLAttribute{v.Nickname, v.Login, v.Username1, v.Username2, v.Email1, v.Email2, v.Email3}
+		for _, c := range candidates {
+			if c != nil && len(c.Values) > 0 && c.Values[0] != nil && c.Values[0].Value != "" {
+				username = c.Values[0].Value
+				break
+			}
+		}
+		if username == "" {
+			return nil, nil
+		}
+		return &extsvc.PublicAccountData{
+			Username: &username,
+		}, nil
```

Comment:
If you remove the nil possibilities, then this becomes:

```suggestion
		v := data.Values

		candidates := []SAMLAttribute{v.Nickname, v.Login, v.Username1, v.Username2, v.Email1, v.Email2, v.Email3}
		for _, c := range candidates {
			if len(c.Values) > 0 && c.Values[0].Value != "" {
				return &extsvc.PublicAccountData{
					Username: &c.Values[0].Value,
				}, nil
			}
		}
		return nil, nil
```

Reply:
There's also a way to do this by unmarshaling it into a `map[string]SAMLAttribute` instead.

But I don't know if one is necessarily better than the other one. It's maybe a bit less clunky? Not sure

Reply:
I think given that we're unmarshaling using a list of candidate strings, it might be better to unmarshal using the map

Reply:
I think I'm going to move this code to the auth provider itself, so that they have a method called `GetExternalAccountData` that will hide all the complexity.